# Homework 5
This homework requires `wine.csv`, and the `tidyverse` and `Rtsne` packages. Install them if you haven't already!  
See the following link for how to add new packages to Binder: https://github.com/rjenki/BIOS512?tab=readme-ov-file#adding-packages-to-installr-later.   
**For readability and easier processing, please make each question part a different code chunk.**

In [1]:
library(tidyverse)
library(Rtsne)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


ERROR: Error in library(Rtsne): there is no package called ‘Rtsne’


## Question 1  

#### a) Import your data.  
#### b) Check out the columns present using one of R's data frame summary.  
#### c) Get summary statistics on the numeric variables.  

In [24]:
library(tidyverse);
wine <- read_csv("wine.csv");
glimpse(wine);
wine %>% pivot_longer(cols = everything(), names_to = "variable", values_to = "value") %>% group_by(variable) %>%
    summarise(mean = mean(value), sd = sd(value), median = median(value)) %>% arrange(desc(mean))
              

Rows: 178 Columns: 14
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (14): Alcohol, Malicacid, Ash, Alcalinity_of_ash, Magnesium, Total_pheno...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Rows: 178
Columns: 14
$ Alcohol                        <dbl> 14.23, 13.20, 13.16, 14.37, 13.24, 14.2…
$ Malicacid                      <dbl> 1.71, 1.78, 2.36, 1.95, 2.59, 1.76, 1.8…
$ Ash                            <dbl> 2.43, 2.14, 2.67, 2.50, 2.87, 2.45, 2.4…
$ Alcalinity_of_ash              <dbl> 15.6, 11.2, 18.6, 16.8, 21.0, 15.2, 14.…
$ Magnesium                      <dbl> 127, 100, 101, 113, 118, 112, 96, 121, …
$ Total_phenols                  <dbl> 2.80, 2.65, 2.80, 3.85, 2.80, 3.27, 2.5…
$ Flavanoids                     <dbl> 3.06, 2.76, 3.24, 3.49, 2.69, 3.39, 2.5…
$ Nonflavanoid_phenols           <dbl> 0.28, 0.26, 0.30, 0.24, 0.39, 0.34, 0.3…
$ Proanthocyanins                <dbl> 2.29, 1.28, 2.81, 2.18, 1.82, 1.97, 1.9…
$ Color_intensity                <dbl> 5.64, 4.38, 5.68, 7.80, 4.32, 6.75, 5.2…
$ Hue                            <dbl> 1.04, 1.05, 1.03, 0.86, 1.04, 1.05, 1.0…
$ `0D280_0D315_of_diluted_wines` <dbl> 3.92, 3.40, 3.17, 3.45, 2.93, 2.85, 3.5…
$ Proline         

variable                     mean        sd          median 
1  Proline                      746.8932584 314.9074743 673.500
2  Magnesium                     99.7415730  14.2824835  98.000
3  Alcalinity_of_ash             19.4949438   3.3395638  19.500
4  Alcohol                       13.0006180   0.8118265  13.050
5  Color_intensity                5.0580899   2.3182859   4.690
6  0D280_0D315_of_diluted_wines   2.6116854   0.7099904   2.780
7  Ash                            2.3665169   0.2743440   2.360
8  Malicacid                      2.3363483   1.1171461   1.865
9  Total_phenols                  2.2951124   0.6258510   2.355
10 Flavanoids                     2.0292697   0.9988587   2.135
11 class                          1.9382022   0.7750350   2.000
12 Proanthocyanins                1.5908989   0.5723589   1.555
13 Hue                            0.9574494   0.2285716   0.965
14 Nonflavanoid_phenols           0.3618539   0.1244533   0.340

## Question 2

#### a) Scale and center your data  
*Hint:* Use a `mutate()` statement across all columns **except class** with `function(x) as.numeric(scale(x))`.

#### b) Based on what you saw in the summary statistic table from the imported data, why would scaling and centering this data be helpful before we perform PCA?


In [26]:
wine %>% select(-class) %>% mutate(across(c(everything()), function (x) as.numeric(scale(x))))


Alcohol     Malicacid   Ash         Alcalinity_of_ash Magnesium  
1    1.51434077 -0.56066822  0.23139979 -1.16630317        1.90852151
2    0.24559683 -0.49800856 -0.82566722 -2.48384052        0.01809398
3    0.19632522  0.02117152  1.10621386 -0.26798225        0.08810981
4    1.68679140 -0.34583508  0.48655389 -0.80697481        0.92829983
5    0.29486844  0.22705328  1.83522559  0.45067448        1.27837900
6    1.47738706 -0.51591132  0.30430096 -1.28607930        0.85828399
7    1.71142720 -0.41744613  0.30430096 -1.46574348       -0.26196936
8    1.30493643 -0.16680747  0.88751034 -0.56742256        1.48842650
9    2.25341491 -0.62332789 -0.71631546 -1.64540766       -0.19195352
10   1.05857838 -0.88291793 -0.35180959 -1.04652705       -0.12193769
11   1.35420804 -0.15785609 -0.24245783 -0.44764644        0.36817315
12   1.37884384 -0.76654998 -0.16955666 -0.80697481       -0.33198519
13   0.92308146 -0.54276546  0.15849862 -1.04652705       -0.75208020
14   2.15487169 -0.54276546  0.08559744 -2.42395246       -0.61204853
15   1.69910930 -0.41744613  0.04914686 -2.24428828        0.15812565
16   0.77526663 -0.47115441  1.21556562 -0.68719868        0.85828399
17   1.60056608 -0.37268923  1.28846679  0.15123418        1.41841067
18   1.02162467 -0.68598755  0.92396093  0.15123418        1.06833150
19   1.46506916 -0.66808479  0.41365272 -0.89680690        0.57822065
20   0.78758453  0.68357369  0.70525741 -1.28607930        1.13834733
21   1.30493643 -0.63227927 -0.31535901 -1.04652705        1.83850567
22  -0.08698653  1.31017034  1.03331269 -0.26798225        0.15812565
23   0.87380985 -0.42639751 -0.02375431 -0.86686287        0.08810981
24  -0.18552975 -0.65913341  0.55945507 -0.50753450       -0.33198519
25   0.61513390 -0.47115441  0.88751034  0.15123418       -0.26196936
26   0.06082829 -0.25632128  3.11099611  1.64843571        1.69847400
27   0.47963697 -0.50695994  0.92396093 -1.01658302       -0.47201686
28   0.36877585 -0.55171684 -0.82566722 -0.74708674       -0.40200103
29   1.07089628 -0.39059199  1.58007149 -0.02843001        0.50820482
30   1.25566482 -0.58752236 -0.57051311 -1.04652705       -0.26196936
⋮   ⋮           ⋮           ⋮           ⋮                 ⋮          
149  0.39341166  0.80889302  0.04914686  0.600394638      -0.5420327 
150  0.09778200  1.39968415 -0.02375431  0.600394638       0.9282998 
151  0.61513390  0.70147646  0.92396093  1.348995406       1.6284582 
152 -0.25943717  0.29866433  0.41365272  0.750114792       0.8582840 
153  0.13473571 -0.39059199  1.39781855  1.798155867       1.1383473 
154  0.28255053  0.86260131 -0.31535901 -0.297926283      -0.1219377 
155 -0.51811312 -0.93662621 -0.97146956  0.151234178       0.2281415 
156  0.20864312  2.55441226 -0.16955666  0.750114792      -0.4720169 
157  1.03394258  1.59661452  0.04914686  0.001514024      -0.7520802 
158 -0.67824585  0.62091403  0.99686210  2.247316327      -0.1919535 
159  1.64983769 -0.58752236  1.21556562  1.648435713      -0.1219377 
160  0.59049809 -0.59647374  0.99686210  0.899834945      -0.7520802 
161 -0.78910697  1.33702448  0.04914686  0.450674485      -0.8220960 
162  0.84917404  0.82679579  0.63235624  0.151234178       0.5082048 
163 -0.18552975  0.83574717  0.77815859  0.750114792       0.4381890 
164 -0.05003283  0.99687202 -0.06020490 -0.297926283       0.4381890 
165  0.96003516  0.37922675 -0.24245783  0.750114792      -0.6820644 
166  0.89844565  1.81144766 -0.38826018  0.899834945      -0.8220960 
167  0.55354439  1.22065654  0.85105976  1.049555099       0.7882682 
168 -0.22248346  0.92526097 -0.24245783  0.001514024      -0.8220960 
169  0.71367712  0.21810190  1.17911504  1.498715559       0.3681732 
170  0.49195487  2.02628080  1.79877500  1.648435713       0.8582840 
171 -0.98619340  0.62091403 -0.16955666 -0.148206130      -0.2619694 
172 -0.28407297  0.04802567 -0.31535901  0.001514024      -0.9621277 
173  1.42811545  0.15544223  0.41365272  0.151234178      -0.6120485 
174  0.8738098

Alcohol Malicacid Ash  Alcalinity_of_ash Magnesium Total_phenols Flavanoids
1   14.23   1.71      2.43 15.6              127       2.80          3.06      
2   13.20   1.78      2.14 11.2              100       2.65          2.76      
3   13.16   2.36      2.67 18.6              101       2.80          3.24      
4   14.37   1.95      2.50 16.8              113       3.85          3.49      
5   13.24   2.59      2.87 21.0              118       2.80          2.69      
6   14.20   1.76      2.45 15.2              112       3.27          3.39      
7   14.39   1.87      2.45 14.6               96       2.50          2.52      
8   14.06   2.15      2.61 17.6              121       2.60          2.51      
9   14.83   1.64      2.17 14.0               97       2.80          2.98      
10  13.86   1.35      2.27 16.0               98       2.98          3.15      
11  14.10   2.16      2.30 18.0              105       2.95          3.32      
12  14.12   1.48      2.32 16.8               95       2.20          2.43      
13  13.75   1.73      2.41 16.0               89       2.60          2.76      
14  14.75   1.73      2.39 11.4               91       3.10          3.69      
15  14.38   1.87      2.38 12.0              102       3.30          3.64      
16  13.63   1.81      2.70 17.2              112       2.85          2.91      
17  14.30   1.92      2.72 20.0              120       2.80          3.14      
18  13.83   1.57      2.62 20.0              115       2.95          3.40      
19  14.19   1.59      2.48 16.5              108       3.30          3.93      
20  13.64   3.10      2.56 15.2              116       2.70          3.03      
21  14.06   1.63      2.28 16.0              126       3.00          3.17      
22  12.93   3.80      2.65 18.6              102       2.41          2.41      
23  13.71   1.86      2.36 16.6              101       2.61          2.88      
24  12.85   1.60      2.52 17.8               95       2.48          2.37      
25  13.50   1.81      2.61 20.0               96       2.53          2.61      
26  13.05   2.05      3.22 25.0              124       2.63          2.68      
27  13.39   1.77      2.62 16.1               93       2.85          2.94      
28  13.30   1.72      2.14 17.0               94       2.40          2.19      
29  13.87   1.90      2.80 19.4              107       2.95          2.97      
30  14.02   1.68      2.21 16.0               96       2.65          2.33      
⋮   ⋮       ⋮         ⋮    ⋮                 ⋮         ⋮             ⋮         
149 13.32   3.24      2.38 21.5               92       1.93          0.76      
150 13.08   3.90      2.36 21.5              113       1.41          1.39      
151 13.50   3.12      2.62 24.0              123       1.40          1.57      
152 12.79   2.67      2.48 22.0              112       1.48          1.36      
153 13.11   1.90      2.75 25.5              116       2.20          1.28      
154 13.23   3.30      2.28 18.5               98       1.80          0.83      
155 12.58   1.29      2.10 20.0              103       1.48          0.58      
156 13.17   5.19      2.32 22.0               93       1.74          0.63      
157 13.84   4.12      2.38 19.5               89       1.80          0.83      
158 12.45   3.03      2.64 27.0               97       1.90          0.58      
159 14.34   1.68      2.70 25.0               98       2.80          1.31      
160 13.48   1.67      2.64 22.5               89       2.60          1.10      
161 12.36   3.83      2.38 21.0               88       2.30          0.92      
162 13.69   3.26      2.54 20.0              107       1.83          0.56      
163 12.85   3.27      2.58 22.0              106       1.65          0.60      
164 12.96   3.45      2.35 18.5              106       1.39          0.70      
165 13.78   2.76      2.30 22.0               90       1.35          0.68      
166 13.73   4.36      2.26 22.5               88       1.28          0.47      
167 

## Question 3

#### a) Perform PCA

#### b) How much of the total variance is explained by PC1? PC2? What function do we use to see that information?

#### c) Why are we doing PCA first?

#### d) What is the rotation matrix? Print it explicitly.  
*Hint:* Check the notes for a simple way to do this!

#### e) Plot PC1 vs. PC2, using the wine class as labels for coloring.  
*Hint:* You'll first need a data set with only PC1 and PC2, then add back the class variable from your scaled data set with a `mutate()` statement. Then, you can use `color = factor(class)` in your `ggplot` statement.

#### f) What do you see after plotting PC1 vs. PC2? What does this mean in context of wine classes?

#### g) Give an example of data where PCA would fail. You can describe the data or do a simulation.  
*Hint:* Our notes have a few examples!

#### h) Explain the difference between vector space and manifold, and how these terms apply to what we did/will do with T-SNE.


## Question 4
#### a) Perform T-SNE
Set `seed = 123`.  
*Hint:* Subset your PCA results to PC1–PC10, add the class variable back in, remove duplicates, then perform T-SNE.

#### b) Plot the results in 2D
*Hint:* Convert your T-SNE results to a tibble and add back the class variable from your scaled data set using a `mutate()` statement. Then, you can use `color = factor(class)` in your `ggplot` statement.

#### c) Why didn't we stop at PCA?


#### d) What other types of data does this workflow make sense for?

